index.js

```diff
@@ -0,0 +1,167 @@
+ * @param  {String} password  Password of the user
+ * @return {Promise}          User object if it gets created or empty object if already exists
+ */
+function createUser(username, password) {
+    return request({
+        json: true,
+        method: 'POST',
+        // eslint-disable-next-line max-len
+        uri: `${sonarHost}/api/users/create?login=${username}&name=${username}&password=${password}`,
+        auth: {
+            username: adminToken
+        }
+    }).catch((err) => {
+        if (err.statusCode === 400 && err.message.includes('already exists')) {
+            return {};
+        }
+
+        throw new Error(`Failed to create user ${username}: ${err.message}`);
+    });
+}
+
+/**
+ * Give specific user push access to the project
+ * @method grantUserPermission
+ * @param  {String} username      Username of the user
+ * @param  {String} projectKey    Unique identifier for the project
+ * @return {Promise}              Nothing if permission granted
+ */
+function grantUserPermission(username, projectKey) {
+    // Always return 204 even with duplicate calls
+    return request({
+        json: true,
+        method: 'POST',
+        // eslint-disable-next-line
+        uri: `${sonarHost}/api/permissions/add_user?login=${username}&permission=scan&projectKey=${projectKey}`,
+        auth: {
+            username: adminToken
+        }
+    }).catch((err) => {
+        throw new Error(`Failed to grant user ${username} permission: ${err.message}`);
+    });
+}
+
+/**
+ * Generate an access token for the given user
+ * @method generateToken
+ * @param  {String} username   username of the user
+ * @return {Promise}           Object with a token field
+ */
+function generateToken(username) {
+    const tokenName = uuidv4();
+
+    return request({
+        json: true,
+        method: 'POST',
+        uri: `${sonarHost}/api/user_tokens/generate?login=${username}&name=${tokenName}`,
+        auth: {
+            username: adminToken
+        }
+    }).catch((err) => {
+        throw new Error(`Failed to generate user ${username} token: ${err.message}`);
+    });
+}
+
+class CoverageSonar extends CoverageBase {
+    /**
+     * Constructor
+     * @method constructor
+     * @param  {Object} config              Configuration object
+     * @param  {String} config.sdApiUrl     URL for Screwdriver API
+     * @param  {String} config.sonarHost    Host for SonarQube Server
+     * @param  {String} config.adminToken   Sonar Admin token
+     */
+    constructor(config) {
+        super();
+
+        this.config = joi.attempt(config, joi.object().keys({
+            sdApiUrl: joi.string().uri().required(),
+            sonarHost: joi.string().uri().required(),
+            adminToken: joi.string().required()
+        }).unknown(true), 'Invalid config for sonar coverage plugin');
+
+        sdApiAuthUrl = `${this.config.sdApiUrl}/v4/coverage/token`;
+        adminToken = this.config.adminToken;
+        sonarHost = this.config.sonarHost;
```

Comment:
Keep in mind that these are static variables. If the class `CoverageSonar` ever gets instantiated more than once with different config values, the later-instantiated objects will overwrite the variables and affect the previous ones.

If this class is only instantiated once then we are fine.

Reply:
Thanks for the notes! it should be created only once during when we start the screwdriver API